server/store/store.go

```diff
@@ -403,72 +408,125 @@ func (s *SQLStore) MattermostToTeamsUserID(userID string) (string, error) {
 }
 
 func (s *SQLStore) GetPostInfoByMSTeamsID(chatID string, postID string) (*storemodels.PostInfo, error) {
-	query := s.getQueryBuilder().Select("mmPostID, msTeamsLastUpdateAt").From("msteamssync_posts").Where(sq.Eq{"msTeamsPostID": postID, "msTeamsChannelID": chatID})
+	tx, err := s.BeginTx()
+	if err != nil {
+		return nil, err
+	}
+
+	var txErr error
+	defer func() {
+		if txErr != nil {
+			if err := s.RollbackTx(tx); err != nil {
+				s.api.LogWarn("Unable to rollback database transaction", "error", err.Error())
+			}
+			return
+		}
+
+		if err := s.CommitTx(tx); err != nil {
+			s.api.LogWarn("Unable to commit database transaction", "error", err.Error())
+		}
+	}()
+
+	query := s.getQueryBuilder().Select("mmPostID, msTeamsLastUpdateAt").From("msteamssync_posts").Where(sq.Eq{"msTeamsPostID": postID, "msTeamsChannelID": chatID}).Suffix("FOR UPDATE").RunWith(tx)
 	row := query.QueryRow()
 	var lastUpdateAt int64
 	postInfo := storemodels.PostInfo{
 		MSTeamsID:      postID,
 		MSTeamsChannel: chatID,
 	}
-	err := row.Scan(&postInfo.MattermostID, &lastUpdateAt)
-	if err != nil {
-		return nil, err
+
+	if txErr = row.Scan(&postInfo.MattermostID, &lastUpdateAt); txErr != nil {
+		return nil, txErr
 	}
 	postInfo.MSTeamsLastUpdateAt = time.UnixMicro(lastUpdateAt)
 	return &postInfo, nil
 }
 
 func (s *SQLStore) GetPostInfoByMattermostID(postID string) (*storemodels.PostInfo, error) {
-	query := s.getQueryBuilder().Select("msTeamsPostID, msTeamsChannelID, msTeamsLastUpdateAt").From("msteamssync_posts").Where(sq.Eq{"mmPostID": postID})
+	tx, err := s.BeginTx()
+	if err != nil {
+		return nil, err
+	}
+
+	var txErr error
+	defer func() {
+		if txErr != nil {
+			if err := s.RollbackTx(tx); err != nil {
+				s.api.LogWarn("Unable to rollback database transaction", "error", err.Error())
+			}
+			return
+		}
+
+		if err := s.CommitTx(tx); err != nil {
+			s.api.LogWarn("Unable to commit database transaction", "error", err.Error())
+		}
+	}()
+
+	query := s.getQueryBuilder().Select("msTeamsPostID, msTeamsChannelID, msTeamsLastUpdateAt").From("msteamssync_posts").Where(sq.Eq{"mmPostID": postID}).Suffix("FOR UPDATE").RunWith(tx)
 	row := query.QueryRow()
 	var lastUpdateAt int64
 	postInfo := storemodels.PostInfo{
 		MattermostID: postID,
 	}
-	err := row.Scan(&postInfo.MSTeamsID, &postInfo.MSTeamsChannel, &lastUpdateAt)
-	if err != nil {
-		return nil, err
+
+	if txErr = row.Scan(&postInfo.MSTeamsID, &postInfo.MSTeamsChannel, &lastUpdateAt); txErr != nil {
+		return nil, txErr
 	}
+
 	postInfo.MSTeamsLastUpdateAt = time.UnixMicro(lastUpdateAt)
 	return &postInfo, nil
 }
 
-func (s *SQLStore) SetPostLastUpdateAtByMattermostID(postID string, lastUpdateAt time.Time) error {
-	query := s.getQueryBuilder().Update("msteamssync_posts").Set("msTeamsLastUpdateAt", lastUpdateAt.UnixMicro()).Where(sq.Eq{"mmPostID": postID})
+func (s *SQLStore) SetPostLastUpdateAtByMattermostID(postID string, lastUpdateAt time.Time, tx *sql.Tx) error {
```

Comment:
This is more a matter of style, I would use the `tx` as the first parameter if we want to pass the `tx` to the store methods. Because my expectation is that every query has the `tx` parameter at the beginning. but that is not the case here. Anyway, I would love to hide all the transaction complexity from the rest of the application. For example, not using transactions where there are not needed, and where they are needed use something special, like a "two steps function", something like a function that starts the operation, and returns a function that finish (or rollback) the operation depending on the parameter passed (maybe something like `err == nil`).